### DISCUSSION TITLE
突然想起 rust 社区点评 zig 的一句话

### DISCUSSION THREAD

**Discussion by kscooo** [zigcc/forum#23] [General]
无意挑起语言之争，之前 rust 社区的一位老哥炮打 golang 的一篇缴文在 hn 反响很大，其中 cue 到 zig 的一句话是：

> 这是一次非常有趣的会议：我仍然为我所做的演讲深感羞愧，但我深情地记得有一次观众问 Go 团队“为什么你选择忽略自 1970 年代以来关于类型系统的任何研究”？我当时并没有完全理解其中的含义，但我现在确实知道了。

> 从那以后，我对我的语言彻底失去了兴趣，因为我开始关心语义而不是语法，这就是为什么我也没有看过 Zig、Nim、Odin 等：我不再对 “a better C” 感兴趣了。

https://fasterthanli.me/articles/lies-we-tell-ourselves-to-keep-using-golang#:~:text=I%20have%20since%20thoroughly%20lost%20interest%20in%20my%20language%2C%20because%20I%27ve%20started%20caring%20about%20semantics%20a%20lot%20more%20than%20syntax%2C%20which%20is%20why%20I%20also%20haven%27t%20looked%20at%20Zig%2C%20Nim%2C%20Odin%2C%20etc%3A%20I%20am%20no%20longer%20interested%20in%20%22a%20better%20C%22.

**Comment by jiacai2050** [comment]
作者是个知名博主，基本上每篇文章都有很大的阅读量，内容深入浅出。

- https://www.reddit.com/r/fasterthanlime/comments/uemda7/comment/i7sawem/?utm_source=share&utm_medium=web2x&context=3

作者在这里有个回复，对 Zig 这些新语言没兴趣的主要原因是没有彻底解决内存安全。

**Reply by kscooo** [reply]
zig team 的人昨天刚跟 rust 社区的几个大佬 battle 内存安全的话题
https://twitter.com/croloris/status/1568342336580755456?s=21&t=v2Dj_F2f_kUzZDQps5KjtQ
https://twitter.com/croloris/status/1568573729940164608?s=21&t=v2Dj_F2f_kUzZDQps5KjtQ
https://twitter.com/croloris/status/1568704125826748416?s=21&t=v2Dj_F2f_kUzZDQps5KjtQ

感觉前几天对线的这波人都还蛮有重量的

**Reply by jiacai2050** [reply]
是不是发生了哪件事导致社区这两天关注度提高了？

PS：我建了个 Twitter List 来跟踪 Zig 社区内 Hacker 的动态
- https://twitter.com/i/lists/1570249876155568129

**Reply by kscooo** [reply]
母鸡，不过貌似导火索是 zig 社区的人引用 tigerbeetle 设计文档的一句话内涵 rust 强调的内存安全，开启后面的对线的。

赞一个，大佬有考虑建立微信群 or slack 这种么，虽然可能就容易变水了，没纯文字贴质量高就是了。

**Reply by jiacai2050** [reply]
- https://zigcc.github.io/
这里有 QQ 、telegram 群，其他的目前没有

**Reply by hhstore** [reply]
- [Loris Cro ](https://twitter.com/croloris) 作为 Zig VP, 发这种带立场(拉踩)的内容, 确实不合适. 
- 他如果是普通开发者, 无所谓. 
- 作为 Zig VP 背景属性, 要求他要做沟通其他社区, 融洽关系, 吸引其他社区开发者的.
- 被喷, 也正常. 属实不应该.
- Zig 和  Rust 不冲突, 可以共存. 
- Zig 应该搞好和 Rust 的互操作性. 毕竟 Rust 生态已经非常强大了.

**Reply by jiacai2050** [reply]
> [Loris Cro ](https://twitter.com/croloris)作为 Zig VP, 发这种带立场(拉踩)的内容, 确实不合适.

上面的 Twitter 看了下，感觉内容还好，没有太强的针对性

**Reply by jiacai2050** [reply]
趁着整理 9 月月报的机会，把这件事的来龙去脉整理了下
- https://zigcc.github.io/monthly/202209/

**Reply by kscooo** [reply]
赞一个，redis 作者也有参与讨论，也有输出对 rust 的一些看法

**Reply by kscooo** [reply]
https://twitter.com/antirez/status/1568624432603701250

**Reply by woodysu** [reply]
所以在Rust看来内存泄漏不算内存安全？ OOM之后程序崩了 效果不是一样？虽然也在用Rust，但是Rust还是比较复杂（很难说和C++最佳实践相比哪个更复杂）， 心智负担还是很重的。Zig是个很好的的选择，但是Zig也应该在不增加太多复杂度的情况下，在内存安全的方向上做一些努力或者提供一些辅助。

**Comment by hhstore** [comment]
## about Go:

- 正巧, 炮打 Go, 我也有一篇. 233
    - https://github.com/hhstore/blog/issues/348
- 看了一下原文, 他对 Go 的不满, 很多是因为他在 Windows 下使用. (笑...
    - 还好我不用 win.
    - Go 的 error 机制, nil 零值, channel 关闭问题等, 都挺恶心. 
    - Go 标准库的质量差, byte 类型等, 还有 Go module 机制. 
    - 总之, 这些黑点, 确实没得洗.
    - Go 只能称之为: 大多数场景下的 a better Java(主要还是成本对比上的). 仅此. 
    - 客观讲: 即使有这么些坑, 踩一遍+掌握, 费时也很短. 比之 Rust/C++, 依然有优势.
- Go 的优势:    
    - 易上手. 难度和 Java, Python 一个 level.
    - 单位成本(硬件一致)下, `性价比最高`(QPS/稳定性/开发效率). 和 Python, Java, PHP, Ruby, Node.js 等对比.
    - 生态规模+成熟度 >> Rust >> Zig. Go 在 Web 中间件领域, 绝对霸权, 这是好事. (当然 Rust 在追赶...
    - fuck Go 之前, 更要 fuck Java, fuck JS, fuck C++.
    - 比讨厌, 还得看 Java, JS, C++.
- Google 家的产品, 设计上基本都比较平庸( Go, Dart), 更多都是为了`省钱(人成本+机器成本)`. 
    - 怎么省钱, 怎么来. 
    - 没有花里胡哨. (所谓的大道至简... = 省钱
    - Google 喜欢/推崇的`方案/技术栈(python, java, go)`, 无一不是省钱(人+机器)的.

## about Rust:

- 对比 Go 的`优势`越来越明显. 
- Go 的困境和 Python 一样. 当思考是写临时代码 vs 写一劳永逸的代码?
- Go 和 Python 都不是`终极方案`(优化终点). 
- 一旦开始思考 `优化终点`, 目光所及只有: `Rust, C/C++, Zig`.

> 终极方案困境: 

- 这个问题, 会魂牵梦绕.
- 这山(Go,Python), 永远没有那山(Rust, C, Zig)高.
- 既然写到死, 还是临时方案(Go/Python). 为何一开始不用 Rust/Zig?
- 那么最终会发现, Go 会比 Python 还要惨, 两头都可取, 中间方案 gg.
- 要写 demo 原型, 用 Python, 反正就是临时, 越快越好. 写好 demo, 直接跳过 go, 改 rust/zig 重写. 结束.



> Rust Web:

- Rust Web 的生态肉眼可见的壮大. 
- Go 社区持续失血, 越来越多的开发者流失. (和当年 Python 一样. 2018 dead 倒计时)
- 大量 `go 中间件`, 在`主动/被动`用 Rust 重写. 案例非常非常多.
- 这个过程, 亦如当年大量 `Java 中间件`, 被 Go 重写掉.



## about Zig: 

- Zig 前路艰难. 
- Zig 的自举?
- Zig 的包管理工具?
- Zig 的生态壮大: 杀手级项目遍地开花?

> 保守估计: 

- 3-5年.
- 亦如 回看2018年的 Rust.

**Reply by t0nghe** [reply]
迷雾通 第 2 版的时候用 Go 写的，现在到了第 4 版，换成了 Rust。

**Reply by hhstore** [reply]
> 迷雾通 第 2 版的时候用 Go 写的，现在到了第 4 版，换成了 Rust。

- 越来越多的go项目, 正在 or 已经 被 rust 重写掉. 
- https://github.com/tidb-incubator/tidis
    - 这是 tidb/tikv 的一个子项目, 最近也是从go 重写成 rust.
    - https://github.com/yongman/tidis
    - 这个趋势非常明显.
